### PR TITLE
Implements ScoreKeeper.

### PR DESCRIPTION
**Pull request by tomsons26**

### CODE REVIEW

**Comment by tomsons26** [comment]
~~dunno what to call int_dword9C_capturedTech and int_dwordA0_capturedscore..~~

File: src/game/common/rts/scorekeeper.cpp [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+
+    xfer->xferInt(&m_playerIndex);
+
+    Xfer_Object_Count_Map(xfer, &m_objectsBuilt);
+
+    uint16_t count = MAX_PLAYER_COUNT;
+
+    xfer->xferUnsignedShort(&count);
+
+    captainslog_relassert(count == MAX_PLAYER_COUNT, 6, "size of objects destroyed array has changed");
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp:63]: CODE_06 or 6? Several times in this file.

File: src/game/common/rts/scorekeeper.cpp [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+
+            auto iter = m_objectsLost.find(tmpl);
+
+            if (iter != m_objectsLost.end()) {
+                count = (*iter).second;
+            }
+
+            int new_count = count + 1;
+            // needs confirming is the right function
+            m_objectsLost[tmpl] = new_count;
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: Shouldn't this be `m_objectsBuild` ?

File: src/game/common/rts/scorekeeper.cpp [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+            auto iter = m_objectsLost.find(tmpl);
+
+            if (iter != m_objectsLost.end()) {
+                count = (*iter).second;
+            }
+
+            int new_count = count - 1;
+            // needs confirming is the right function
+            m_objectsLost[tmpl] = new_count;
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: Shouldn't this be `m_objectsBuild` ?

File: src/game/common/rts/scorekeeper.cpp [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+
+void ScoreKeeper::Xfer_Object_Count_Map(Xfer *xfer, ScoreKeeperCountMap *count_map)
+{
+    captainslog_relassert(count_map, 6, "Invalid map parameter");
+
+    uint8_t version = 1;
+    xfer->xferVersion(&version, 1);
+
+    // i think?
+    uint16_t count = (uint16_t)(*count_map).size();
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: `static_cast<uint16_t>(count_map->size())`

File: src/game/common/rts/scorekeeper.cpp [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+    xfer->xferUnsignedShort(&count);
+
+    Utf8String name;
+
+    if (xfer->Get_Mode() == 1) {
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: `XFER_SAVE`

File: src/game/common/rts/scorekeeper.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+#include "bitflags.h"
+#include "gametype.h"
+#include "snapshot.h"
+#include <map>
+
+class Object;
+class ThingTemplate;
+
+typedef std::map<const ThingTemplate *, int> ScoreKeeperCountMap;
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.h]: `int32_t` because Xfer

File: src/game/common/rts/scorekeeper.cpp [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+        for (auto &it : *count_map) {
+            name = it.first->Get_Name();
+            xfer->xferAsciiString(&name);
+            int c = it.second;
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: `int32_t value`

File: src/game/common/rts/scorekeeper.cpp [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+            xfer->xferInt(&c);
+        }
+    } else {
+        for (int i = 0; i < count; ++i) {
+            xfer->xferAsciiString(&name);
+            ThingTemplate *tmpl = g_theThingFactory->Find_Template(name, true);
+
+            captainslog_relassert(tmpl != nullptr, 6, "Unknown thing template '%s'", name.Str());
+
+            int c;
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: `int32_t value`

File: src/game/common/rts/scorekeeper.h [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    int m_totalMoneyEarned;
+    int m_totalMoneySpent;
+
+    int m_totalUnitsDestroyed[MAX_PLAYER_COUNT];
+    int m_totalUnitsBuilt;
+    int m_totalUnitsLost;
+
+    int m_totalBuildingsDestroyed[MAX_PLAYER_COUNT];
+    int m_totalBuildingsBuilt;
+    int m_totalBuildingsLost;
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.h]: `int32_t` because Xfer

File: src/game/common/rts/scorekeeper.cpp [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+    xfer->xferVersion(&version, 1);
+
+    // i think?
+    uint16_t count = (uint16_t)(*count_map).size();
+    xfer->xferUnsignedShort(&count);
+
+    Utf8String name;
+
+    if (xfer->Get_Mode() == 1) {
+        for (auto &it : *count_map) {
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: `auto &pair`

File: src/game/common/rts/scorekeeper.cpp [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+            name = it.first->Get_Name();
+            xfer->xferAsciiString(&name);
+            int c = it.second;
+            xfer->xferInt(&c);
+        }
+    } else {
+        for (int i = 0; i < count; ++i) {
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: `uint16_t i`

File: src/game/common/rts/scorekeeper.cpp [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+
+    Xfer_Object_Count_Map(xfer, &m_objectsBuilt);
+
+    uint16_t count = MAX_PLAYER_COUNT;
+
+    xfer->xferUnsignedShort(&count);
+
+    captainslog_relassert(count == MAX_PLAYER_COUNT, 6, "size of objects destroyed array has changed");
+
+    for (int i = 0; i < count; ++i) {
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: `uint16_t i`

File: src/hooker/setuphooks_zh.cpp [hunk 13]
```diff
@@ -2225,4 +2225,21 @@ void Setup_Hooks()
     Hook_Any(0x004E7FB0, ProjectileStreamUpdate::Friend_New_Module_Instance);
     Hook_Any(0x0061C410, ProjectileStreamUpdate::Add_Projectile);
     Hook_Any(0x0061C650, ProjectileStreamUpdate::Set_Position);
+
+    // scorekeeper.cpp
```
Comment (xezon) [inline src/hooker/setuphooks_zh.cpp:2239]: `#include "scorekeeper.h"`

File: src/game/common/rts/scorekeeper.cpp [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+    if (g_theGameLogic->Get_Enable_Scoring()) {
+        bool add_to_map = false;
+        const ThingTemplate *tmpl = object->Get_Template();
+
+        if (tmpl->Is_KindOf(KINDOF_STRUCTURE)) {
+
+            if (tmpl->Is_KindOf(KINDOF_SCORE)) {
+                ++int_dwordA0_capturedscore;
+            } else {
+                ++int_dword9C_capturedTech;
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: I tested this in game. It triggers when completing capture of civilian building. On some unofficial map it also triggered when beginning to capture a civilian building, which made little sense.

File: src/game/common/rts/scorekeeper.cpp [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+void ScoreKeeper::Add_Object_Captured(const Object *object)
+{
+    if (g_theGameLogic->Get_Enable_Scoring()) {
+        bool add_to_map = false;
+        const ThingTemplate *tmpl = object->Get_Template();
+
+        if (tmpl->Is_KindOf(KINDOF_STRUCTURE)) {
+
+            if (tmpl->Is_KindOf(KINDOF_SCORE)) {
+                ++int_dwordA0_capturedscore;
```
Comment (xezon) [inline src/game/common/rts/scorekeeper.cpp]: I tested this in game. It triggers when completing capture of enemy player building.

**Review by xezon — APPROVED** [review]
Logically looks good.

**Pull request merged**